### DISCUSSION TITLE
Old text shown when it was wider than new text

### DISCUSSION THREAD

**Discussion by ferni7** [mathoudebine/turing-smart-screen-python#110] [General]
Hi All,

I've been playing with the turing smart screen and this python library (thanks for this!).

I used the simple-program.py as example code and written my own function to pull metrics from my Prometheus server where I have things like Room Temperatures and Appliance power usage to display on here. All working ok.

The issue I'm seeing is that when a value goes from say 180W to 989W, the last half of the old W from the width of 180W is till shown.

What is the correct way to fix this so it doesn't show the old text when the old text was wider than the current text?

![image](https://user-images.githubusercontent.com/15047029/208321952-c73fba33-cfb5-4adc-9d90-0e05a8ab1d84.png)

**Comment by mathoudebine** [comment]
Hi, unfortunately this is a known limitation because the new text is drawn over the old one (to save some time by not erasing it first).
For some texts that vary in size, the shorter text does not completely cover the old one. This makes some "ghosting" like in your picture.

Hopefully there are some ways to fix it like we made for the system monitor:
- Add some empty "padding" characters before/after your new text to be sure it always covers the old text when updated. This can be done with:
```python
text_watt=f"{your_numeric_watt_val:>5}W"    # Text is aligned to the right, 5 chars of padding are added to the left
```
- There may still be some issues depending on your font: classic fonts do not have the same width for all characters. Even with padding the new text size might still have a width that vary depending on what you write. If you still have issue after adding padding, select a monospaced font. You have some in res/fonts/ (the ones with "mono" in the name")

**Reply by ftk789** [reply]
At least is there a way to erase that specific text then re-draw it? I know it takes time, But it's better than being redrawn, especially if the text is changing and can be longer/shorter etc... Would be nice if there's a way to erase it then re-draw it. @mathoudebine

**Reply by mathoudebine** [reply]
It may be a good feature for the future, at the moment there is no easy way to do it because the size of the previous text is not backed up so when a new text has to be displayed there is no way to get the previous text size to erase it first.
There would be also a "blink effect" where the old text would disappear leaving the box empty for a short time before the new text is displayed. This has to be taken into account

**Reply by ftk789** [reply]
> It may be a good feature for the future, at the moment there is no easy way to do it because the size of the previous text is not backed up so when a new text has to be displayed there is no way to get the previous text size to erase it first. There would be also a "blink effect" where the old text would disappear leaving the box empty for a short time before the new text is displayed. This has to be taken into account

I agree, It would still be awesome though, I was thinking of actually detecting where that text is in the background, Then crop the background according to the text size and where it is, Then create an image of the background but cropped to where the text was, Then add your own text, A bit complicated, But worth noting, I wonder how the CPU percentage changes and get's smaller / bigger and doesn't leave a trace behind, I wonder if there's a " live text " option or something, Makes me wonder a lot.

**Reply by mathoudebine** [reply]
For the CPU percentage (and all other sensors that have numeric values), the text has a fixed width that match the max value (e.g. `100%` so 4 characters wide). If the value to display has a smaller width some spaces are added as padding before the value (so 3% will be displayed as `  3%`). This way the old text is always erased either by new digits or by spaces which will display the background.

**Reply by ftk789** [reply]
WOW! That totally makes sense, But pretty weird not erasing the old state of the text and draw the new one, But wow that's really interesting, Well thank you for your explanation, From that Now I know it's not possible at the moment unless a feature get's created.

**Comment by ferni7** [comment]
Thank you @mathoudebine switching to mono font and right aligned text worked :)
